cmd/query/app/handler.go

```diff
@@ -0,0 +1,312 @@
+)
+
+// HTTPHandler handles http requests
+type HTTPHandler interface {
+	RegisterRoutes(router *mux.Router)
+}
+
+type structuredResponse struct {
+	Data   interface{}       `json:"data"`
+	Total  int               `json:"total"`
+	Limit  int               `json:"limit"`
+	Offset int               `json:"offset"`
+	Errors []structuredError `json:"errors"`
+}
+
+type structuredError struct {
+	Code    int        `json:"code,omitempty"`
+	Msg     string     `json:"msg"`
+	TraceID ui.TraceID `json:"traceID,omitempty"`
+}
+
+// APIHandler implements the query service public API by registering routes at httpPrefix
+type APIHandler struct {
+	spanReader       spanstore.Reader
+	dependencyReader dependencystore.Reader
+	adjuster         adjuster.Adjuster
+	logger           zap.Logger
+	queryParser      queryParser
+	httpPrefix       string
+}
+
+// NewAPIHandler returns an APIHandler
+// TODO use Builder to allow this package to move to github while accepting custom components like HAProxyMerge()
+func NewAPIHandler(spanReader spanstore.Reader, dependencyReader dependencystore.Reader, logger zap.Logger, httpPrefix string, adjusters []adjuster.Adjuster) *APIHandler {
+	return &APIHandler{
+		spanReader:       spanReader,
+		dependencyReader: dependencyReader,
+		logger:           logger,
+		queryParser: queryParser{
+			traceQueryLookbackDuration: defaultTraceQueryLookbackDuration,
+		},
+		adjuster:   adjuster.Sequence(adjusters...),
+		httpPrefix: httpPrefix,
+	}
+}
+
+// RegisterRoutes registers routes for this handler on the given router
+func (aH *APIHandler) RegisterRoutes(router *mux.Router) {
+	// TODO why single-trace resource below accept POST method?
+	router.HandleFunc(fmt.Sprintf("/%s/traces/{%s}", aH.httpPrefix, traceIDParam), aH.getTrace).Methods(http.MethodPost, http.MethodGet)
+	router.HandleFunc(fmt.Sprintf(`/%s/traces`, aH.httpPrefix), aH.search).Methods(http.MethodGet)
+	router.HandleFunc(fmt.Sprintf(`/%s/services`, aH.httpPrefix), aH.getServices).Methods(http.MethodGet)
+	// TOOD - remove this when UI catches up
+	router.HandleFunc(fmt.Sprintf("/%s/services/operations", aH.httpPrefix), aH.getOperations).Methods(http.MethodGet)
+	router.HandleFunc(fmt.Sprintf("/api/services/{%s}/operations", serviceParam), aH.getOperationsLegacy).Methods(http.MethodGet)
+	router.HandleFunc(fmt.Sprintf("/%s/dependencies", aH.httpPrefix), aH.dependencies).Methods(http.MethodGet)
+}
+
+func (aH *APIHandler) getServices(w http.ResponseWriter, r *http.Request) {
+	services, err := aH.spanReader.GetServices()
+	if aH.handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	structuredRes := structuredResponse{
+		Data:  services,
+		Total: len(services),
+	}
+	aH.writeJSON(w, &structuredRes)
+}
+
+func (aH *APIHandler) getOperationsLegacy(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	service := vars[serviceParam] //given how getOperationsLegacy is used, this will never work never be empty
```

Comment:
Legacy is the function that is called for endpoints like:

api/services/{someServiceName}/operations.

We agreed to decommission this kind of operation name lookup endpoint.